python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+        """
+        Make a histogram of the distribution of pixel values for each amp.
+
+        Main image data histogram plotted in blue. Edge pixels, if masked, are
+        in red. Note that masked edge pixels do not contribute to the
+        underflow and overflow numbers.
+
+        Parameters
+        ----------
+        pdfPages: `matplotlib.backends.backend_pdf.PdfPages`
+            If not None, append per-amp histograms to this object
+        """
+        fig, ax = plt.subplots(nrows=4, ncols=4, sharex='col', sharey='row', figsize=(13, 10))
+
+        detector = exp.getDetector()
+        expTime = exp.getInfo().getVisitInfo().getExposureTime()
+
+        for (amp, a) in zip(reversed(detector), ax.flatten()):
+            mi = exp.maskedImage[amp.getBBox()]
+
+            # normalise by expTime as we plot in ADU/s and don't always work with master calibs
+            mi.image.array /= expTime
+            stats = afwMath.makeStatistics(mi, afwMath.MEANCLIP | afwMath.STDEVCLIP)
+            mean, sigma = stats.getValue(afwMath.MEANCLIP), stats.getValue(afwMath.STDEVCLIP)
+
+            # Get array of pixels
+            EDGEBIT = exp.maskedImage.mask.getPlaneBitMask("EDGE")
+            imgData = mi.image.array[(mi.mask.array & EDGEBIT) == 0].flatten()
+            edgeData = mi.image.array[(mi.mask.array & EDGEBIT) != 0].flatten()
+
+            thrUpper = mean + nSigmaUsed*sigma
+            thrLower = mean - nSigmaUsed*sigma
+
+            nRight = len(imgData[imgData > thrUpper])
+            nLeft = len(imgData[imgData < thrLower])
+
+            nsig = 6
+            leftEdge = mean - nsig * nSigmaUsed*sigma
+            rightEdge = mean + nsig * nSigmaUsed*sigma
+            nbins = np.linspace(leftEdge, rightEdge, 1000)
+            ey, bin_borders, patches = a.hist(edgeData, histtype='step', bins=nbins, lw=1, edgecolor='red')
+            y, bin_borders, patches = a.hist(imgData, histtype='step', bins=nbins, lw=3, edgecolor='blue')
+
+            # Report number of entries in over-and -underflow bins, i.e. off the edges of the histogam
+            nOverflow = len(imgData[imgData > rightEdge])
+            nUnderflow = len(imgData[imgData < leftEdge])
+
+            # Put v-lines and textboxes in
+            a.axvline(thrUpper, c='k')
+            a.axvline(thrLower, c='k')
+            msg = f"{amp.getName()}\nmean:{mean: .1f}\n$\sigma$:{sigma: .1f}"  # noqa: W605
+            a.text(0.65, 0.6, msg, transform=a.transAxes, fontsize=11)
+            msg = f"nLeft:{nLeft}\nnRight:{nRight}\nnOverflow:{nOverflow}\nnUnderflow:{nUnderflow}"
+            a.text(0.03, 0.6, msg, transform=a.transAxes, fontsize=11.5)
+
+            # set axis limits and scales
+            a.set_ylim([1., 1.7*np.max(y)])
+            lPlot, rPlot = a.get_xlim()
+            a.set_xlim(np.array([lPlot, rPlot]))
+            a.set_yscale('log')
+            a.set_xlabel("ADU/s")
+
+        return
+
+    def _setEdgeBits(self, exposureOrMaskedImage):
```

Comment:
Please move this above the plotting methods, since it's more connected with the code higher up.